### ISSUE TITLE
Changes needed for Anki 2.1.24

### ISSUE THREAD

**Issue by dae** [hssm/advanced-browser#70]
Hi Houssam,

Just wanted to give you a heads up about some fairly big changes in 2.1.24 that will affect some of your add-ons. The bad news is that it is no longer possible to define database functions in Python, as the DB layer is now handled in Rust, so the approach you were taking to sort some fields will no longer be possible. I have made some changes that will hopefully minimise the disruption however:

- The ignore accents functionality is now integrated into Anki, and much faster
- Something similar to your valueForField() db function is integrated into Anki
- Sorting on fields like decks is now integrated into Anki
- There are some new hooks in the browse screen to make it easier to override the search order, and you could use a similar approach to what Anki's using with a temp table for complex sorts: https://github.com/ankitects/anki/blob/cc44523449cda621c20f80e6583040deb57debb1/rslib/src/search/cards.rs#L71

I've made the minimal changes to the advanced browser add-on to get it to start, and sorting on user fields seems to work. I haven't given it a full test, and action will be required to get some of the sorting+caching going again, but hopefully this is a start.

Patches here: https://github.com/ankitects/advanced-browser/commits/2.1.24-fixes
And change notes here: https://betas.ankiweb.net/#/?id=changes-in-2124beta1

Hope you're going well with all the recent craziness.

**Comment by Arthur-Milchior** [comment]
Please, @hssm, let me know if you want help. As I've passed time on this add-on recently, I hope I already know it

**Comment by hssm** [comment]
Hi Damien!

Thanks for doing the work to get this add-on running again, and thanks for all the work you continue to do on Anki. I'm excited to see development picking up again.

Sorry for the delay in getting back to this. It's taken some time for me to catch up on how things work again (and for me to remember how my own add-on is supposed to work). There's a lot to digest in that set of commits but I think I'm starting to get it.

There are some discrepancies with the old add-on and the new one. I've made a list. Maybe we can address some of them:

**Empty values appear on top**
Older version made sure empty values are at the end. New one does not:
![image](https://user-images.githubusercontent.com/789082/77757331-27dd5780-7085-11ea-895d-b7312b9ba294.png)
The old approach relied on the temporary table and [a particular pattern](https://github.com/hssm/advanced-browser/blob/master/advancedbrowser/advancedbrowser/core.py#L239) to sort them last. I don't think I can repeat this easily with the current query. I'd like to, if possible, so I'll continue to experiment with it.

SQLite now has a `NULLS LAST` keyword that can be added to the `order by` clause, and I [tried adding it](https://github.com/hssm/advanced-browser/blob/master/advancedbrowser/advancedbrowser/note_fields.py#L109), but it doesn't seem to work. This might be because the empty field value is being [returned as a string](https://github.com/ankitects/anki/blob/cc44523449cda621c20f80e6583040deb57debb1/rslib/src/storage/sqlite.rs#L77) instead of null. Unfortunately I don't know enough (any) rust to experiment with this function. (It would have to convert valid, empty fields to nulls as well).


**Media/HTML is not stripped**
![image](https://user-images.githubusercontent.com/789082/77758480-2745c080-7087-11ea-82fc-563dff1db68c.png)
Columns with HTML or media filenames don't sort properly. Old version was using `anki.utils.stripHTMLMedia` on the field value. I *think* the equivalent version is in rust now? Can we make it use it somehow?

**Caching**
I haven't taken any before/after measurements, but it seems relatively quick even without the temporary table. I might be willing to drop it. It would simplify the add-on a lot, so I don't mind it. *Unless* it was the media stripping that was slowing it down, then we may need it again.

**Two broken columns**
- Sorting of "current deck" column (is this different from the built-in `Original Deck` column? @Arthur-Milchior )
- Sorting of the "overdue interval" column

**Note-only mode**
Not being activated in the new function. I think it's easy to add back in.

--

In any case, I've released a new version of the add-on on AnkiWeb targeting 2.1.24, just to get things going for people who rely on it.

@Arthur-Milchior, thank you very much for your offer to help. If there's anything on this list that you can improve that could help.

**Comment by Arthur-Milchior** [comment]
"Original deck" is empty when the card is not in a filtered deck. "Current deck" shows the filtered deck of the card if it is filtered, and the unfiltered deck otherwise. So yeah, those columns are different.

**Comment by hssm** [comment]
Sorry I meant the original `Deck` column. There are three deck columns now:
![image](https://user-images.githubusercontent.com/789082/77784148-2d9b6300-70ae-11ea-8f86-b344240e09e4.png)

The one this add-on creates shows the same info as the built-in `Deck` but without the original deck it came from in brackets. The built-in one isn't sorting by deck name though.

--

With regards to the ignore accents add-on, I'm glad that there's a faster approach now since the speed was one of the complaints it had. Since it's considerably faster now, does it make sense to turn it on permanently for everyone?

**Comment by Arthur-Milchior** [comment]
Can't we use https://docs.rs/rusqlite/0.21.0/rusqlite/functions/index.html in anki in order to still define function. (I don't know whether rust has easy access to deck name. Clearly in this case, a table sending did to deck name would be the most useful value to have)

**Comment by Arthur-Milchior** [comment]
The name "Current deck (filtered)" is lying. The point of "Current name" is that it shows the current value and only this. That's usually what I want to see myself. And if you eventually accept the "edit table content" then that's also what we need in order to allow this simple edition. Because as soon as there are two different values, it can't be easily parsed anymore

**Comment by Arthur-Milchior** [comment]
Concerning the note mode: Please see https://ankiweb.net/shared/info/797076357
It's currently compatible with advanced browser.
Thanks to recent hooks, it can easily be created as a separate add-on. So people can install extra columns and/or note mode separately. I suggest that, for the sake of code simplicity, the note mode should be removed from Advanced Browser (potentially, add a link in the readme).
I can also remove the code myself if you want.

**Comment by dae** [comment]
> **Empty values appear on top**
> Older version made sure empty values are at the end. New one does not:
> ![image](https://user-images.githubusercontent.com/789082/77757331-27dd5780-7085-11ea-895d-b7312b9ba294.png)
> The old approach relied on the temporary table and [a particular pattern](https://github.com/hssm/advanced-browser/blob/master/advancedbrowser/advancedbrowser/core.py#L239) to sort them last.

That seems complicated, why not something like 'val = "" asc, val asc' in the order by clause.

> SQLite now has a `NULLS LAST` keyword that can be added to the `order by` clause, and I [tried adding it](https://github.com/hssm/advanced-browser/blob/master/advancedbrowser/advancedbrowser/note_fields.py#L109), but it doesn't seem to work.

If it's the 'note type has no such field' case you're referring to, perhaps false needs to be changed to NULL there, or you need to use the same trick as above.

> **Media/HTML is not stripped**
> ![image](https://user-images.githubusercontent.com/789082/77758480-2745c080-7087-11ea-82fc-563dff1db68c.png)
> Columns with HTML or media filenames don't sort properly. Old version was using `anki.utils.stripHTMLMedia` on the field value. I _think_ the equivalent version is in rust now? Can we make it use it somehow?
> 
> **Caching**
> I haven't taken any before/after measurements, but it seems relatively quick even without the temporary table. I might be willing to drop it. It would simplify the add-on a lot, so I don't mind it. _Unless_ it was the media stripping that was slowing it down, then we may need it again.

The nc search is not the default because it's 2-3x slower. HTML stripping is 4x slower than an nc search, and can make the browser quite laggy with larger collections, so I don't think it should be done on the fly - it's the main reason why stock Anki requires the user to select a particular sort field, so that the stripping can be done once then used many times.

One possibility in the future would be to pack a stripped version of all fields into the sort field, but that would break the numeric sorting case, and some benchmarking would need to be done before going down that path.

> @dae Can't we use https://docs.rs/rusqlite/0.21.0/rusqlite/functions/index.html in anki in order to still define function. 

Yes, that is how Anki is providing field_at_index and regexp. But database functions that depend on external state are bad design - it is better to restructure the queries to use a temporary table to do any lookups that are necessary. I linked an example of that above (https://github.com/ankitects/anki/blob/cc44523449cda621c20f80e6583040deb57debb1/rslib/src/search/cards.rs#L71), and you could accomplish the same thing in Python by writing the table yourself, then passing in a custom order by clause to use it.

**Comment by hssm** [comment]
I've managed to restore all of the initial characteristics of the add-on. That includes html/media stripping and (attempted) numerical sorting for any note field column.

I used the temporary table approach and it adds a lot of overhead. So yeah, it's slow, but I'm working with what I have. I did find an easy optimization by pre-compiling some regexes which halved the runtime. The performance is about the same as pre-2.1.24 (without the optimization, as in the released version), some columns faster and some slower, but overall it's close to before so I'm reasonably content.

I've uploaded a new release targeting 2.1.24+ to AnkiWeb with all of these fixes.
